framework/sandstone.cpp

```diff
@@ -2685,6 +2709,118 @@ static void warn_deprecated_opt(const char *opt)
+    return true;//TO DO implement idle trigger for Windows
+    #endif
+
+    return false;
+}
+
+static void background_scan_init(void)
+{
+    MonotonicTimePoint now = MonotonicTimePoint::clock::now();
+
+    // init timestamps to 1day old - this quickstarts testing on first run
+    if ((sApp->background_scan.init_timestamp == true) && (sApp->service_background_scan == true)) {
+        sApp->background_scan.timestamp.fill(now - sApp->background_scan.time_to_run_next_batch_of_tests);
+    }
+}
+
+static void background_scan_update_load_threshold(void)
+{
+    MonotonicTimePoint now = MonotonicTimePoint::clock::now();
+
+    hours time_from_last_test = 
+        duration_cast<hours>(now - sApp->background_scan.timestamp[sApp->background_scan.timestamp_newest]);
+    // scale our idle threshold value from 0.2 base, to 0.8 after 12h
+    // every hour adds 0.05 to the threshold value
+    sApp->background_scan.load_idle_threshold = 
+        sApp->background_scan.load_idle_threshold_init +
+        (time_from_last_test.count() * sApp->background_scan.load_idle_threshold_inc_val);
+
+    // prevent the idle threshold form rising above 0.8
+    if(sApp->background_scan.load_idle_threshold > sApp->background_scan.load_idle_threshold_max)
+        sApp->background_scan.load_idle_threshold = sApp->background_scan.load_idle_threshold_max;
+}
+
+static void background_scan_wait(void) 
+{
+    MonotonicTimePoint now = MonotonicTimePoint::clock::now();
+    // Don't run tests unless load is low or it's time to run a test anyway
+    while(1) {
+        // wait ~5 mins no matter what
+        sApp->delay_between_tests = sApp->background_scan.minimum_delay_between_tests;
+
+        double wait_deviation_percent = 10.0;
+
+        wait_delay_between_tests_with_deviation(
+                wait_deviation_percent,
+                false);
+
+        now = MonotonicTimePoint::clock::now();
+
+        // If all the last 24 tests ran in less than the last 24h, don't run
+        // any tests at all
+        if (now < (sApp->background_scan.timestamp[sApp->background_scan.timestamp_newest] + sApp->background_scan.time_to_run_next_batch_of_tests)) {
+            sApp->delay_between_tests = sApp->background_scan.time_to_run_next_batch_of_tests;
+
+            double wait_deviation_percent = 0.1;
+            wait_delay_between_tests_with_deviation(
+                    wait_deviation_percent,
+                    false);
+        }
+
+        background_scan_update_load_threshold();
+
+        // if the system is idle, run a test
+        if (system_is_idle(sApp->background_scan.load_idle_threshold))
+            break;
+
+        // if we haven't run *any* tests in the last x hours, run a test
+        // because of day/night cycles, 12 hours should help typical data center
+        // duty cycles.
+        if (now > (sApp->background_scan.timestamp[sApp->background_scan.timestamp_newest] + sApp->background_scan.time_to_force_next_test_running))
```

Comment:
Hint: for the `static constexpr` variables, you can write `SandstoneBackgroundScan::time_to_force_next_test_running`